### ISSUE TITLE
Use standard deviation from global explanation graph to calculate standard deviation of prediction?

### ISSUE THREAD

**Issue by bverhoeff** [interpretml/interpret#235]
(moved this from other issue)
First of all many thanks for creating and maintaining this project.

I have the following related questions about quantifying uncertainty of predictions:

1) Is a standard deviation calculated by bagging data 8 times and training 8 models and averaging them representative for the distribution of uncertainty of the feature scores of a certain feature at certain feature values? (e.g. by calculating a normal distribution with average and SD).

2) Can we 'propagate' this uncertainty towards the final predictions by drawing from the respective abovementioned distributions as apropriate multiple times and calculating predictions multiple times as well, finally resulting in an average and SD of predictions for a certain case.

3) Currently there are no upper and lower bounds for interaction terms in the internal object. Would it be difficult to add these? Otherwise abovementioned method would only be possible without interaction terms

4) Most importantly, would abovementioned method be mathematically correct? I can imagine that this is the wrong place for that question. I could ask this at Stats Exchange / Cross Validated?

code example to illustrate line of thought:

```python
def feature_mean_sd(ebm_global, feature: int, value: float):
    name = ebm_global.__dict__["_internal_obj"]["overall"]["names"][feature]
    if " x " in name:
        return 0, 0, name  # return zeros when interaction, not supported

    feat_dict = ebm_global.__dict__["_internal_obj"]["specific"][feature]
    vals = feat_dict["names"]
    means = feat_dict["scores"]
    upps = feat_dict["upper_bounds"]

    i = np.searchsorted(vals, value)
    if i == len(vals) - 1:  # in case new value above range X_train
        i -= 1
    sd = upps[i] - means[i]

    return means[i], sd, name


def draw_feature_value(ebm_global, feature, value):
    mean, sd, name = feature_mean_sd(ebm_global, feature, value)

    return np.random.normal(mean, sd), name


def pred_with_sd(ebm_global, feature_values):
    samples = []

    for sample in range(16):
        pred = intercept
        for f in range(len(feature_values)):
            pred += draw_feature_value(ebm_global, f, feature_values[f])[0]
        samples.append(pred)
        
    return np.mean(samples), np.std(samples)


ebm_global = ebm.explain_global(name='EBM')
intercept = ebm.intercept_[0]

samples_range, samples_sd = [], []
for i in range(0, 100):
    feature_values = X_test.iloc[i, :].values
    samples = pred_with_sd(ebm_global, feature_values)
    print(f"samples: mean = {samples[0]:.3f}, sd = {samples[1]:.3f}, y_true = {y_test[i]}")
    samples_range.append(samples[0])
    samples_sd.append(samples[1])

print(f"Lowest mean: {min(samples_range):.3f}, highest mean: {max(samples_range):.3f}")

import matplotlib.pyplot as plt
plt.scatter(samples_range, samples_sd, c=y_test[:100] + 1, alpha=0.5)
plt.show()
```

Results with KDD2015 dropout data (64 columns selected using lightgbm's feature importance):

...
samples: mean = 3.099, sd = 0.051, y_true = True
samples: mean = 3.038, sd = 0.100, y_true = True
samples: mean = 2.603, sd = 0.079, y_true = True
samples: mean = 3.169, sd = 0.056, y_true = True
samples: mean = 0.413, sd = 0.623, y_true = False
samples: mean = -1.751, sd = 0.220, y_true = False
samples: mean = 1.611, sd = 0.165, y_true = True
samples: mean = 2.808, sd = 0.101, y_true = True
samples: mean = 3.055, sd = 0.091, y_true = True
samples: mean = 2.153, sd = 0.157, y_true = True
Lowest mean: -3.615, highest mean: 3.289

mean vs SD, lower prediction score means higher uncertainty. Could be right given the lower number of cases:
![image](https://user-images.githubusercontent.com/39224153/118413708-fec0c180-b6a0-11eb-916c-cf76b87087ef.png)

Last question: where is the function for the final transformation stored?

**Comment by interpret-ml** [comment]
Hi @bverhoeff, 

This is a great topic!

- _Is a standard deviation calculated by bagging data 8 times and training 8 models and averaging them (as you did) representative for the distribution of uncertainty of the feature scores of a certain feature at certain feature values? (e.g. so that I can draw samples from that distribution?)._

Unfortunately, uncertainty estimates derived by bagging are not perfect confidence intervals. Bagging can underestimate the true variance in uncertainty of feature scores, especially in regions with relatively small sample size. However, we believe it can give reasonable hints about how the model might behave under shifts in the training data distribution. I'd highly recommend bumping up the number of bags (controlled by the `outer_bags` parameter) to something like 30 or 50 if you can afford the extra computational time, to get better estimates of this bagged uncertainty. We set it to 8 by default as a reasonable tradeoff between compute speed, accuracy, and intelligibility, but for more serious applications increasing this number is recommended.

- _Currently there are no upper and lower bounds for interaction terms in the internal object. Would it be difficult to add these?_

Good point! We left these off the internal object because it isn't clear how to visualize uncertainty estimates of interaction terms. However, it's easy for us to calculate them and include them in the data structures, even if they don't get visualized today. We'll be sure to include this in a future release!

- _If we draw samples from each feature graphs distribution for a certain values for a case and use them to calculate a prediction; and if we would repeat this a number of times, could we determine the average+-SD of the prediction for that case as well? And would that be a correct way to determine uncertainty of the predictions? ...would abovementioned method be mathematically correct?_

This is tricky question! I believe this requires making fairly strong assumptions about the uncertainty distribution on each learned value. It's probably reasonable to assume it is Gaussian with the mean and standard deviations stored by the model, but I'm not completely sure about this. The other (possibly bigger) concern is that this method implicitly assumes independence of error distributions across features by drawing samples from each feature independently, which might lead to miscalibrated estimates. 

One other option we're thinking about is allowing users to keep each of the bagged models as a property of the EBMs. That way you can empirically derive prediction intervals by using each model to make a prediction and calculating a range. This is very close to the approach you were suggesting, but doesn't make the independence or distribution assumptions -- in this case, we'd have access to each of the base models to generate samples directly. You could then do something like this:

```python3
def pred_from_base_models(ebm, instances):
    binned_inst = ebm.preprocessor_.transform(instances)
    
    preds_per_bag = np.zeros(shape=(instances.shape[0], ebm.outer_bags))
    for bag_index, core_ebm in enumerate(ebm.bagged_models_):
        output = np.zeros(shape=(instances.shape[0]))
        for i in range(instances.shape[0]):
            term_contribs = [None] * instances.shape[1]
            for feature, bin_index in enumerate(binned_inst[i]):
                term_contribs[feature] = core_ebm.model_[feature][bin_index]

            output[i] = np.sum(term_contribs)
        preds_per_bag[:, bag_index] = output
        
    # Calc mean and stdev
    return np.c_[np.mean(preds_per_bag, axis=1), np.std(preds_per_bag, axis=1)]
```

The downsides are that this would make the models much larger (we throw away each mini bag currently), and would also be significantly slower than the current prediction pipeline. However, it might be reasonable to include as an option for users who are willing to pay that cost. 

I ran some experiments comparing prediction intervals derived through both methods on the Adult income dataset. Here's the distribution of prediction intervals for both methods, where the blue trace ("uncertainties") is the approach you mentioned, and the red trace ("base preds") is where the intervals are calculated with each of the base models. 

![image](https://user-images.githubusercontent.com/50031429/118846416-ad7c2200-b89a-11eb-8dc1-83bd4a68e9ba.png)

In this case, they're fairly close! However, when testing on another medical dataset, the differences become more apparent:

![image](https://user-images.githubusercontent.com/50031429/118846994-4317b180-b89b-11eb-8865-89acf4d9c24a.png)

It's not clear if either approach is perfect, but we believe option 2 tends to give better estimates (due to the fewer assumptions being made). Would this be useful for you? We're still actively researching this, and would love to keep the discussion going! 

-InterpretML Team

**Comment by bverhoeff** [comment]
While looking at the forest I didn't see the tree! Obviously your solution looks way more promising both visually and conceptually. I wonder how we could validate that. Would you know about a correct bayesian method to estimate standard deviations of the predictions so they can be compared with the bagging results? Would be a nice paper!

By the way: standard errors of interactions would still be interesting, especially for local explanations. E.g. the horizontal bar chart local explanation might be more informative with error bars for each feature.

**Comment by onacrame** [comment]
A probabilistic estimation would be of immense value.  I'm in financial services, where people in the business are always wary of point estimates and ask the question, "how confident are we of this prediction?".  We've experimented with quantile regression within the usual GBDT packages (highly inefficient as you can only model two quantiles at a time) and also looked at NGBoost, which is not particularly fast.

**Comment by rodrigovssp** [comment]
#up

**Comment by rodrigovssp** [comment]
#agremiaoinfantil
#agremiaojuvenil

**Comment by paulbkoch** [comment]
@bverhoeff & @onacrame -- As of release v0.3.0, we now store the individual bagged models inside the EBM under the bagged_scores_ attribute, so it should now be possible to implement the pred_from_base_models function above.  I will add this to our backlog since exposing a built-in version would be useful to a wider audience.

**Comment by joshdunnlime** [comment]
> A probabilistic estimation would be of immense value. I'm in financial services, where people in the business are always wary of point estimates and ask the question, "how confident are we of this prediction?". We've experimented with quantile regression within the usual GBDT packages (highly inefficient as you can only model two quantiles at a time) and also looked at NGBoost, which is not particularly fast.

Here are some examples of other packages doing distribution learning. These can be used to get any quantile.

XGBoost-Distribution: Like NGBoost - but faster: https://github.com/CDonnerer/xgboost-distribution
LightGBMLSS - https://github.com/StatMixedML/LightGBMLSS
Also an XGBoostLLS - https://github.com/StatMixedML/XGBoostLSS

Additionally, XGBoost and CatBoost now support multi quantile predictions.

EBM would really benefit from [custom loss](https://github.com/interpretml/interpret/issues/281), or even better, out-of-the-box multi pinball loss 😍

**Comment by paulbkoch** [comment]
In addition to storing the individual bagged models, we now have a [predict_with_uncertainty](https://interpret.ml/docs/python/api/ExplainableBoostingClassifier.html#interpret.glassbox.ExplainableBoostingClassifier.predict_with_uncertainty) function .  [Pinball loss](https://github.com/interpretml/interpret/blob/main/shared/libebm/compute/objectives/PinballRegressionObjective.hpp) is also implemented and will be available in the next release.
